File: Kotlin/ExtratoController.kt
package com.kamino.banking.controller

import com.kamino.banking.entity.*
import com.kamino.banking.repository.*
import com.kamino.banking.service.ExtratoService
import com.kamino.banking.service.RetornoHelper
import jakarta.persistence.EntityManager
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.math.BigDecimal
import java.time.LocalDateTime
import java.util.*

// ==================== Controller ====================

@RestController
@RequestMapping("/api/banking/kamino")
@CrossOrigin(origins = ["*"])
class ExtratoController(
    val extratoService: ExtratoService,
    val transacaoRepository: TransacaoFinanceiraRepository,
    val entityManager: EntityManager
) {
    private val logger = LoggerFactory.getLogger(ExtratoController::class.java)

    private val requestCache = HashMap<String, Any>()

    @PostMapping("/transactions/reload/hook")
    fun transactionsReloadHook(
        @RequestParam(required = false, defaultValue = "false") forcar: Boolean,
        @RequestHeader("X-Company-Token") companyToken: String
    ): RetornoHelper {
        logger.info("Atualizando pagamentos por hook")
        return extratoService.atualizarContasBancarias(companyToken, forcar)
    }

    @GetMapping("/transactions/pending")
    fun getTransacoesPendentes(
        @RequestParam companyName: String,
        @RequestParam idContaBanco: Int
    ): Any {
        val cacheKey = "pending_$companyName_$idContaBanco"
        if (requestCache.containsKey(cacheKey)) {
            return requestCache[cacheKey]!!
        }
        val result = extratoService.consultarPendentes(companyName, idContaBanco)
        if (result != null) {
            requestCache[cacheKey] = result
        }
        return result ?: emptyList<Any>()
    }

    @GetMapping("/balance/pending")
    fun getSaldoPendente(
        @RequestParam companyName: String,
        @RequestParam idContaBanco: Int
    ): Any {
        return extratoService.getSaldoPendente(companyName, idContaBanco)
            ?: mapOf("balance" to 0)
    }

    @GetMapping("/extrato")
    fun consultarExtrato(
        @RequestParam idContaBanco: Int,
        @RequestParam dataInicio: String,
        @RequestParam dataFim: String,
        @RequestParam(required = false, defaultValue = "0") pagina: Int
    ): ResponseEntity<Any> {
        val inicio = LocalDateTime.parse(dataInicio)
        val fim = LocalDateTime.parse(dataFim)
        
        val transacoes = transacaoRepository.findByContaBancoAndPeriodo(idContaBanco, inicio, fim)
        return ResponseEntity.ok(transacoes)
    }

    @PostMapping("/contas/atualizar")
    fun atualizarContas(@RequestBody body: Map<String, Any>): RetornoHelper {
        val companyToken = body["companyToken"] as String
        val forcar = body["forcar"] as? Boolean ?: false
        return extratoService.atualizarContasBancarias(companyToken, forcar)
    }

    @DeleteMapping("/cache/limpar")
    fun limparCache(): Map<String, Any> {
        requestCache.clear()
        return mapOf("sucesso" to true, "mensagem" to "Cache limpo")
    }

    @GetMapping("/relatorio/saldo-diario")
    fun getRelatórioSaldoDiario(
        @RequestParam idContaBanco: Int,
        @RequestParam dataInicio: String,
        @RequestParam dataFim: String
    ): List<Map<String, Any>> {
        val query = """
            SELECT data, SUM(valor) as saldo 
            FROM extrato_banco 
            WHERE id_conta_banco = $idContaBanco 
            AND data BETWEEN '$dataInicio' AND '$dataFim'
            GROUP BY data 
            ORDER BY data
        """
        val result = entityManager.createNativeQuery(query).resultList
        return result.map { row ->
            val arr = row as Array<*>
            mapOf("data" to arr[0], "saldo" to arr[1])
        }
    }
}

File: TypeScript/extrato.controller.ts
import { Controller, Get, Post, Delete, Query, Body, Headers, Logger } from '@nestjs/common';
import { DataSource } from 'typeorm';
import { ExtratoService, RetornoHelper } from './extrato.service';
import { TransacaoFinanceiraRepository } from './repositories';

@Controller('api/banking/kamino')
export class ExtratoController {
  private readonly logger = new Logger(ExtratoController.name);
  private requestCache: Record<string, any> = {};

  constructor(
    private readonly extratoService: ExtratoService,
    private readonly transacaoRepository: TransacaoFinanceiraRepository,
    private readonly dataSource: DataSource,
  ) {}

  @Post('transactions/reload/hook')
  async transactionsReloadHook(
    @Query('forcar') forcar: string = 'false',
    @Headers('X-Company-Token') companyToken: string,
  ): Promise<RetornoHelper> {
    this.logger.log('Atualizando pagamentos por hook');
    return this.extratoService.atualizarContasBancarias(companyToken, forcar === 'true');
  }

  @Get('transactions/pending')
  async getTransacoesPendentes(
    @Query('companyName') companyName: string,
    @Query('idContaBanco') idContaBanco: string,
  ): Promise<any> {
    const cacheKey = `pending_${companyName}_${idContaBanco}`;
    if (this.requestCache[cacheKey]) {
      return this.requestCache[cacheKey];
    }

    const result = await this.extratoService.consultarPendentes(companyName, parseInt(idContaBanco));
    if (result) {
      this.requestCache[cacheKey] = result;
    }

    return result ?? [];
  }

  @Get('balance/pending')
  async getSaldoPendente(
    @Query('companyName') companyName: string,
    @Query('idContaBanco') idContaBanco: string,
  ): Promise<any> {
    const result = await this.extratoService.getSaldoPendente(companyName, parseInt(idContaBanco));
    return result ?? { balance: 0 };
  }

  @Get('extrato')
  async consultarExtrato(
    @Query('idContaBanco') idContaBanco: string,
    @Query('dataInicio') dataInicio: string,
    @Query('dataFim') dataFim: string,
    @Query('pagina') pagina: string = '0',
  ): Promise<any> {
    const inicio = new Date(dataInicio);
    const fim = new Date(dataFim);

    const transacoes = await this.transacaoRepository.findByContaBancoAndPeriodo(
      parseInt(idContaBanco),
      inicio,
      fim,
    );
    return transacoes;
  }

  @Post('contas/atualizar')
  async atualizarContas(@Body() body: any): Promise<RetornoHelper> {
    const companyToken = body.companyToken;
    const forcar = body.forcar ?? false;
    return this.extratoService.atualizarContasBancarias(companyToken, forcar);
  }

  @Delete('cache/limpar')
  limparCache(): any {
    this.requestCache = {};
    return { sucesso: true, mensagem: 'Cache limpo' };
  }

  @Get('relatorio/saldo-diario')
  async getRelatorioSaldoDiario(
    @Query('idContaBanco') idContaBanco: string,
    @Query('dataInicio') dataInicio: string,
    @Query('dataFim') dataFim: string,
  ): Promise<any[]> {
    const query = `
      SELECT data, SUM(valor) as saldo
      FROM extrato_banco
      WHERE id_conta_banco = ${idContaBanco}
      AND data BETWEEN '${dataInicio}' AND '${dataFim}'
      GROUP BY data
      ORDER BY data
    `;

    const result = await this.dataSource.query(query);
    return result;
  }
}
